gh-136134: Fallback to next auth method when CRAM-MD5 fails due to unsupported hash (e.g. FIPS)

nogueira-raphael opened the pull request:
This change improves the `smtplib.SMTP.login()` method to handle the case where CRAM-MD5 fails due to missing or unsupported MD5 digest (e.g. in FIPS-compliant environments). Instead of crashing with a `ValueError`, the client now skips CRAM-MD5 and tries the next available mechanism.

A regression test has been added to verify this fallback behavior.

nogueira-raphael committed:
"gh-136134: Fallback to next auth method when CRAM-MD5 fails due to unsupported hash (e.g. FIPS)

nogueira-raphael committed:
gh-136134: Add code comment explaining CRAM-MD5 fallback and NEWS entry

kannansasientrust commented on Lib/smtplib.py:
```diff
@@ -743,6 +743,10 @@ def login(self, user, password, *, initial_response_ok=True):
                 last_exception = e
+            except ValueError as e:
+                last_exception = e
+                if 'unsupported' in str(e).lower():
```
why do we care about this specific error? shouldn't the next method be tried regardless of the nature of the error?

nogueira-raphael replied:
In general, it makes sense to fallback regardless of the error. However, in this case, we handle this specific ValueError to deal with environments where FIPS mode is enabled, which disables certain hashing algorithms like MD5.

In FIPS mode, CRAM-MD5 will raise a ValueError from the underlying OpenSSL implementation (e.g. [digital envelope routines] unsupported). Since this isn't a misconfiguration or a generic programming error, but rather an expected, reproducible environment-specific issue, we explicitly catch it and fallback silently to the next available method (e.g. LOGIN).

kannansasientrust replied:
I see your point. I had seen the error only on Linux container with a specific version of openssl. My worry is that the fix will be rendered ineffective if openssl changes the error string or returns a different error on a different OS/platform.

nogueira-raphael replied:
I don't think so, we are saving the error, and try another method, if none of them resolve, we will return the latest error.
However, I also don't like this kind of fix, I didn't found another better way to solve this.
And the error is not only related with that docker image.

picnixz replied:
We should do it differently. Relying on the content of an exception message is a bad idea as it can evolve in the future. Instead, locate where we call HMAC and wrap that call, then raise an appropriate exception and catch that one instead.

If the call is too deep in the stack we should find another way.

nogueira-raphael replied:
Thanks for the suggestion @picnixz, I’ve applied the proposed approach by wrapping the HMAC call directly and raising a dedicated SMTPAuthHashUnsupportedError instead of relying on the exception message. I agree with this approach.

If there’s anything you’d like to adjust in naming or behavior, I’ll be happy to apply it.

nogueira-raphael committed:
bpo-136134: Raise SMTPAuthHashUnsupportedError when HMAC fails in CRAM-MD5

Wraps the HMAC call in auth_cram_md5 and raises a dedicated exception
to avoid relying on error message content. Also updates the relevant tests.

kannansasientrust commented on Lib/smtplib.py:
```diff
@@ -743,13 +750,12 @@ def login(self, user, password, *, initial_response_ok=True):
-            except ValueError as e:
+            except SMTPAuthHashUnsupportedError as e:
                 # Some environments (e.g., FIPS) disable certain hashing algorithms like MD5,
                 # which are required by CRAM-MD5. This raises a ValueError when trying to use HMAC.
```
name of the exception in the comment needs update

nogueira-raphael replied:
Good catch, tks 
I’ve updated the comment to reflect the correct exception name

nogueira-raphael committed:
bpo-136134: Clarify comment on handling unsupported hash in CRAM-MD5

picnixz commented on Lib/smtplib.py:
```diff
@@ -665,8 +669,11 @@ def auth_cram_md5(self, challenge=None):
+            return self.user + " " + hmac.HMAC(
+                self.password.encode('ascii'), challenge, 'md5').hexdigest()
+        except ValueError as e:
+            raise SMTPAuthHashUnsupportedError(f'CRAM-MD5 failed: {e}') from e
```
Separate all calls and only wrap the hmac.HMAC call in a try-catch. TiA

picnixz commented on Lib/test/test_smtplib.py:
```diff
@@ -1570,5 +1570,60 @@ def testAUTH_PLAIN_initial_response_auth(self):
 
 
+class TestSMTPLoginValueError(unittest.TestCase):
+    def broken_hmac(*args, **kwargs):
```
Make it an external function, and mock the entire class instead.

picnixz commented on Lib/smtplib.py:
```diff
@@ -743,6 +750,12 @@ def login(self, user, password, *, initial_response_ok=True):
             except SMTPAuthenticationError as e:
                 last_exception = e
+            except SMTPAuthHashUnsupportedError as e:
+                # Some environments (e.g., FIPS) disable certain hashing algorithms like MD5,
```
This comment is too verbose. Just write something like "skip unsupported HMAC-HASH algorithms"

picnixz reviewed (commented):
I would like to review it in 10 days when I am on my latpop. Meanwhile I can only offer simple comments
